### PR TITLE
Enable all ddp params for hpu parallel strategy

### PR DESCRIPTION
**Pull request by jerome-habana**
Signed-off-by: Jerome <janand@habana.ai>

## What does this PR do?

Expose all available ddp params to be used through hpu parallel strategy

Fixes #\<issue_number>

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/accelerators/test_hpu.py [hunk 1]
```diff
@@ -240,3 +240,31 @@ def test_hpu_auto_device_count():
...
+    )
+    assert isinstance(trainer.strategy, HPUParallelStrategy)
+
+    if trainer.strategy._ddp_kwargs["bucket_cap_mb"] != _bucket_cap_mb:
+        assert False, "bucket_cap_mb not matching !"
+    if trainer.strategy._ddp_kwargs["gradient_as_bucket_view"] != _gradient_as_bucket_view:
+        assert False, "gradient_as_bucket_view not matching !"
+    if trainer.strategy._ddp_kwargs["static_graph"] != _static_graph:
+        assert False, "static_graph not matching !"
+    if trainer.strategy._ddp_kwargs["find_unused_parameters"] != _find_unused_parameters:
```
Comment (kaushikb11) [inline tests/accelerators/test_hpu.py:269]: You can directly add assert to the condition

File: tests/accelerators/test_hpu.py [hunk 2]
```diff
@@ -240,3 +240,31 @@ def test_hpu_auto_device_count():
 def test_hpu_unsupported_device_type():
     with pytest.raises(MisconfigurationException, match="`devices` for `HPUAccelerator` must be int, string or None."):
         Trainer(accelerator="hpu", devices=[1])
+
+
+@RunIf(hpu=True)
+def test_strategy_params_with_hpu_parallel_plugin():
```
Comment (kaushikb11) [inline tests/accelerators/test_hpu.py:246]:
```suggestion
def test_strategy_params_with_hpu_parallel_strategy():
```

**Review by kaushikb11 — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: tests/accelerators/test_hpu.py [hunk 3]
```diff
@@ -240,3 +240,31 @@ def test_hpu_auto_device_count():
 def test_hpu_unsupported_device_type():
     with pytest.raises(MisconfigurationException, match="`devices` for `HPUAccelerator` must be int, string or None."):
         Trainer(accelerator="hpu", devices=[1])
+
+
+@RunIf(hpu=True)
+def test_strategy_params_with_hpu_parallel_strategy():
+    _bucket_cap_mb = 100
```
Comment (carmocca) [inline tests/accelerators/test_hpu.py:247]: Local variables don't need to be prefixed by `_`

File: tests/accelerators/test_hpu.py [hunk 4]
```diff
@@ -240,3 +240,31 @@ def test_hpu_auto_device_count():
...
+    _gradient_as_bucket_view = True
+    _static_graph = True
+    _find_unused_parameters = True
+    trainer = Trainer(
```
Comment (carmocca) [inline tests/accelerators/test_hpu.py:251]: I don't think this test needs to instantiate a `Trainer`, you can do

`strategy = HPUParallelStrategy(...)`

and use that variable for the assertions

File: pytorch_lightning/strategies/hpu_parallel.py [hunk 5]
```diff
@@ -77,7 +90,7 @@ def setup_environment(self) -> None:
     def determine_ddp_device_ids(self) -> None:
         return None
 
-    def pre_configure_ddp(self):  # type: ignore
+    def _pre_configure_ddp(self):  # type: ignore
```
Comment (carmocca) [inline pytorch_lightning/strategies/hpu_parallel.py:93]: Why prefix it with `_`? Now it doesn't match the name in `ddp.py`

File: pytorch_lightning/strategies/hpu_parallel.py [hunk 6]
```diff
@@ -47,9 +48,15 @@ def __init__(
...
         parallel_devices: Optional[List[torch.device]] = None,
+        cluster_environment: Optional[ClusterEnvironment] = None,
         checkpoint_io: Optional[CheckpointIO] = None,
         precision_plugin: Optional[PrecisionPlugin] = None,
+        ddp_comm_state: Optional[object] = None,
+        ddp_comm_hook: Optional[callable] = None,  # type: ignore
+        ddp_comm_wrapper: Optional[callable] = None,  # type: ignore
+        model_averaging_period: Optional[int] = None,
         process_group_backend: Optional[str] = "hccl",
+        **kwargs: Union[Any, Dict[str, Any]],
```
Comment (carmocca) [inline pytorch_lightning/strategies/hpu_parallel.py:59]: The convention is this:

```suggestion
        **kwargs: Any,
```

File: pytorch_lightning/strategies/hpu_parallel.py [hunk 7]
```diff
@@ -47,9 +48,15 @@ def __init__(
         self,
         accelerator: Optional["pl.accelerators.accelerator.Accelerator"] = None,
         parallel_devices: Optional[List[torch.device]] = None,
+        cluster_environment: Optional[ClusterEnvironment] = None,
         checkpoint_io: Optional[CheckpointIO] = None,
         precision_plugin: Optional[PrecisionPlugin] = None,
+        ddp_comm_state: Optional[object] = None,
+        ddp_comm_hook: Optional[callable] = None,  # type: ignore
+        ddp_comm_wrapper: Optional[callable] = None,  # type: ignore
```
Comment (carmocca) [inline pytorch_lightning/strategies/hpu_parallel.py:56]: These callables should be `typing.Callable` and then you shouldn't need that `type: ignore`

File: pytorch_lightning/strategies/hpu_parallel.py (hunk 5 above)
Comment (jerome-habana) [reply pytorch_lightning/strategies/hpu_parallel.py:93]: I would like to keep it with the child and use it for only torch where static_graph is not a ddp param

**Review by carmocca — APPROVED** [review]

**Comment by kaushikb11** [comment]
Approved

**Review by kaushikb11 — APPROVED** [review]
